### PR TITLE
[file-search] update quick open file sort order

### LINKED ISSUE #4348: [file-search] Prefer file name over path when boosting the search results
Use case:
 - Search for `langu`.
 - I would like to see all TS modules starting with `langu` ranked higher than other files with matches in the path.

Theia (with Gitpod):
screen shot 2019-02-14 at 19 13 52

VS Code:
screen shot 2019-02-14 at 19 21 01

### PR DESCRIPTION
**Pull request by vince-fugnitto**
Fixes #4348

Favor sorting quick-open file results by `Name` over the display of `URI`.

**Algorithm**:
1. Favor highest `fuzzy` label ranking.
2. (If above step is equal) Favor lowest substring index.
3. (If above step is equal) Favor shortest label length (indicates a better match label length/query length).
4. (If above step is equal) Fallback to alphabetical order.
5. Repeat above steps using `path`.

<div align='center'>

screen shot 2019-02-19 at 7 45 42 pm

</div>

---

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
@kittaakos is this more what you had in mind when sorting the quick file search?

**Comment by kittaakos** [comment]
@svenefftinge, do you remember why did you switch from `fuzzySearch: true` to `false`?

https://github.com/theia-ide/theia/commit/79f56f142956c7fbdf378b993bf49d24db27b935#diff-4252eba404c6d1d1402a32959247ec83L94

**Comment by svenefftinge** [comment]
Don't remember. This comment sounds like it was matching against stuff that is not even visible.
https://github.com/theia-ide/theia/pull/2542#discussion_r209208935

**Comment by svenefftinge** [comment]
So I tried, and the changes break the ordering so that recently opened come first.

**Comment by svenefftinge** [comment]
screenshot 2019-02-18 at 10 13 59

**Comment by vince-fugnitto** [comment]
thanks @svenefftinge for pointing it out, I'll try and see what I can do to correct the sorting

**Comment by vince-fugnitto** [comment]
@kittaakos @elaihau I updated the PR to not break the existing `recently opened` group, while sorting the `file results` by name and then by the path if needed. 

Can you please re-test whenever you get a chance ? 😄

**Comment by kittaakos** [comment]
I am reviwing it now...

Please check the failing tests on Linux and macOS.

**Comment by kittaakos** [comment]
Thank you for the update, @vince-fugnitto.

I think this is incomplete:
 - First off, I have opened three `package.json` files (from `filesystem`, `file-search`, and `process`), they're were not among the `recently opened` items. I had to reopen the files multiple times to see them in the search results.
 - Unfortunately, I have different results than in VS Code.

Search for `packa`, I can see the `README.md` from `packages/bunyan` but not the other `package.json` files.

VS Code:
screen shot 2019-02-19 at 09 46 20

Theia in Gitpod:
screen shot 2019-02-19 at 09 47 43

screen shot 2019-02-19 at 09 45 56



screen shot 2019-02-19 at 09 45 50

- Finally, as I see, you have added some special handling if the file name starts with the query term, but it does not work when there is not an exact match. For example, when I type `ackage` instead of `package`.

screen shot 2019-02-19 at 09 51 59
screen shot 2019-02-19 at 09 51 45


Is there a chance to do the same as VS Code does in the [`openFileHandler`](https://github.com/Microsoft/vscode/blob/cb281cc758a9fab36a3b6be8ef28bdad74133598/src/vs/workbench/contrib/search/browser/openFileHandler.ts#L137-L150), so that the user has the same search experience?

**Comment by vince-fugnitto** [comment]
> * First off, I have opened three `package.json` files (from `filesystem`, `file-search`, and `process`), they're were not among the `recently opened` items. I had to reopen the files multiple times to see them in the search results.

I noticed the issue in `master` as well.
What's odd is that it seems to work fine for small workspaces, but for large ones the bug is more noticeable.

> * Unfortunately, I have different results than in VS Code.

I'll see what I can do to try and align the experience more with vscode :)

**Comment by vince-fugnitto** [comment]
@kittaakos I updated the PR and improved the custom sort function to more closely resemble that of `vscode`, would you mind giving it another test? :)

File: packages/file-search/src/browser/quick-file-open.ts [hunk 1]
```diff
@@ -192,6 +203,79 @@ export class QuickFileOpenService implements QuickOpenModel, QuickOpenHandler {
...
+        member: 'getLabel' | 'getUri' = 'getLabel'): number {
+
+        /**
+         * Compare two numbers.
+         *
+         * @param x number for comparison.
+         * @param y number for comparison.
+         */
+        function cmp(x: number, y: number): number {
+            return x === y ? 0 : (x > y ? 1 : -1);
```
Comment (kittaakos) [inline packages/file-search/src/browser/quick-file-open.ts]:
```ts
return x - y;
```
?

**Comment by kittaakos** [comment]
One thing I have noticed, when I press `Ctrl/Cmd`+`Shift`+`P` without typing anything, no files are shown.

screen shot 2019-02-20 at 10 13 23

**Comment by kittaakos** [comment]
Freely speaking; the previous version was better, I believe. When I search for `packag` I have the following results.

screen shot 2019-02-20 at 10 16 31

**Comment by vince-fugnitto** [comment]
> One thing I have noticed, when I press `Ctrl/Cmd`+`Shift`+`P` without typing anything, no files are shown.
> 
> screen shot 2019-02-20 at 10 13 23

I did notice the same behaviour in master. By default without any search query only the 'recently opened' items are displayed. It does take a while and some tricky manipulation to actually register an editor as 'recently opened' which may be the culprit and should be fixed in a following PR.

**Comment by vince-fugnitto** [comment]
> Freely speaking; the previous version was better, I believe. When I search for `packag` I have the following results.
> 
> screen shot 2019-02-20 at 10 16 31

For the same query string `packag` I get the following result:

screenshot from 2019-02-20 06-48-11

**Comment by vince-fugnitto** [comment]
I found a way to get the `fuzzy score` so I'll try using that way if we get more reliable results :smile:

**Comment by vince-fugnitto** [comment]
@kittaakos @elaihau I updated the pr to improve the sort algorithm and I believe it is much better now :) 
I described my algorithm here https://github.com/theia-ide/theia/pull/4350#issue-253202757.

**Comment by kittaakos** [comment]
@vince-fugnitto, it works very well. 👍

Can you please try out your own PR in Gitpod, (or with any other setup where the backend is not on the same host and you have to go through the wire), so you can check the user experience.

I think the search performance has dropped a lot. And for instance, when I open the *Command Palette* for the file search, I cannot see any files unless I type at least one character. The `Recently Opened` are also not always there. These all could be related to the latency.

(I am trying your changes locally too...)

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto, it works very well. 
> 
> Can you please try out your own PR in Gitpod, (or with any other setup where the backend is not on the same host and you have to go through the wire), so you can check the user experience.
> 
> I think the search performance has dropped a lot. And for instance, when I open the _Command Palette_ for the file search, I cannot see any files unless I type at least one character. The `Recently Opened` are also not always there. These all could be related to the latency.
> 
> (I am trying your changes locally too...)

Sure I'll give it a try!

**Comment by vince-fugnitto** [comment]
@kittaakos I see the recently opened items and the performance seems instantaneous to me :(

screenshot from 2019-02-20 11-32-37

**Comment by vince-fugnitto** [comment]
> Thank you for the patch!

thank you for your help and feedback!

**Pull request merged**